packages/vm/src/memory.rs

```diff
@@ -86,7 +91,7 @@ pub fn write_region(ctx: &Ctx, ptr: u32, data: &[u8]) -> Result<usize, Error> {
 }
 
 // Reads in a ptr to Region in wasm memory and constructs the object we can use to access it
-fn to_region(ctx: &Ctx, ptr: u32) -> Region {
+pub fn to_region(ctx: &Ctx, ptr: u32) -> Region {
```

Comment:
I think to write a region into wasm would be something like this: (probably buggy, but to explain my idea):

```rust
pub fn update_wasm_region(ctx: &Ctx, ptr: u32, region: Region) {
    let memory = ctx.memory(0);
    let wptr = WasmPtr::<Region>::new(ptr);
    let cell = wptr.deref(memory).unwrap();
    cell.set(region);
}
```

But I'm not sure if that is truly safe with memory layouts (works for read). Alternately, we only need to update length, right? And can just write a u32, 8 bytes after ptr.

Reply:
Assuming

```rust
#[repr(C)]
struct Region {
```

has the same memory layout in the Wasm contract and the Rust VM is something we do all the time. If that is not safe, we are in big trouble.

But assuming the u32 fields in `Region` start with 4 bytes offset from each other is dangerous. There are some advanced rules that define the layout in memory, optimized for access speed. The rules are documented here: https://doc.rust-lang.org/std/mem/fn.size_of.html#size-of-structs It turns out that our struct is 12 bytes long. But I would not assume future readers know the memory layout.